client/app/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/client/Camera1Controller.java

```diff
@@ -161,43 +164,40 @@ public void onPictureTaken(byte[] data, Camera camera) {
-//
-//            ByteBuffer byteBuffer = ByteBuffer.allocate(image.getByteCount());
-//            image.copyPixelsToBuffer(byteBuffer);
-//
-//            int width = image.getWidth();
-//            int height = image.getHeight();
-//
-//            byte[] rgbData = byteBuffer.array();
-//            for (int input = 0, output = 0; output < 4 * width * height; output++) {
-//                rgbData[input++] = rgbData[output++];  // red
-//                rgbData[input++] = rgbData[output++];  // green
-//                rgbData[input++] = rgbData[output++];  // blue
-//            }
-//
-//            RGBImageData rgb = new RGBImageData(rgbData, width, height);
-
-            callback.onPreviewFrame(new RGBImageData(data,
-                    previewSize.getWidth(), previewSize.getHeight()));
+            callback.onPreviewFrame(data, previewSize);
         }
     };
 
     /**
-     * Choose the best preview size from a list of options given a target
+     * Choose the best preview size from a list of options given an aspect ratio
      * Note that target and return assume portrait display while options assumes landscape
-     * @param target Dimension representing the target size for the preview (in portrait form)
+     * @param aspectRatio Dimension representing the aspect ratio that the preview should be
      * @param options A List of Camera.Size options for the size of the preview (in landscape form)
      * @return The best size picked from the list (in portrait form)
      */
-    private Dimension getBestPreviewSize(Dimension target, List<Camera.Size> options) {
-        //TODO pick the best size
-        // for now just pick the first size
-        // switch width and height to convert between portrait and landscape
-        int width = options.get(0).height;
-        int height = options.get(0).width;
-        return new Dimension(width, height);
+    private Dimension getBestPreviewSize(Dimension aspectRatio, List<Camera.Size> options) {
+        // pick the smallest preview size that maintains the aspect ratio
+        // initially sort by size ascending
+        Collections.sort(options, new Comparator<Camera.Size>() {
+            @Override
+            public int compare(Camera.Size lhs, Camera.Size rhs) {
+                return (lhs.width * lhs.height) - (rhs.width * rhs.height);
+            }
+        });
+
+        // pick the first that matches the aspect ratio given
+        // have to swap width and height
+        double targetRatio = (double) aspectRatio.getHeight() / (double) aspectRatio.getWidth();
+        for(Camera.Size size : options) {
+            double sizeRatio = (double) size.width / (double) size.height;
+            if(targetRatio == sizeRatio) {
+                return new Dimension(size.width, size.height);
+            }
+        }
+
+        // if we didn't find a size print an error and return null
+        // TODO handle this gracefully
+        throw new NoSupportedPreviewSizeException();
```

Comment:
We assume the preview fills the entire screen, because if it doesn't this creates scaling problems that we currently can't deal with. Hence we require that the preview is the same aspect ratio as the screen.